File: iOSChatExample/iOSChatExample/Models/Message.swift
import Foundation

enum MessageType {
    case Sender
    case Receiver
    case Common
}

struct Message: Identifiable, Equatable, Hashable {
    var participant: String?
    var text: String
    var id = UUID()
    var contentType: String
    var messageType: MessageType
    var timeStamp: String
    var messageID: String?
    var status: String?
    var isRead: Bool = false

    var content: MessageContent? {
        switch contentType {
        case ContentType.plainText.rawValue:
            return PlainTextContent.decode(from: text)
        case ContentType.richText.rawValue:
            // A rich text content class could be created later as complexity increse
            return PlainTextContent.decode(from: text)
        case ContentType.interactiveText.rawValue:
            return decodeInteractiveContent(from: text)
        default:
            // Handle or log unsupported content types
            return nil
        }
    }
    
    // Helper method to decode interactive content
    private func decodeInteractiveContent(from text: String) -> MessageContent? {
        guard let jsonData = text.data(using: .utf8),
              let genericTemplate = try? JSONDecoder().decode(GenericInteractiveTemplate.self, from: jsonData) else {
            return nil
        }
        switch genericTemplate.templateType {
            case QuickReplyContent.templateType:
                return QuickReplyContent.decode(from: text)
            case ListPickerContent.templateType:
                return ListPickerContent.decode(from: text)
            // Add cases for each interactive message type, decoding as appropriate.
            default:
                print("Unsupported interactive content type: \(genericTemplate.templateType)")
                return nil
        }
    }
}

File: mobileChatExamples/androidChatExample/app/src/main/java/com/blitz/androidchatexample/models/Message.kt
package com.blitz.androidchatexample.models

import com.blitz.androidchatexample.utils.ContentType
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import java.util.UUID

enum class MessageType{
    SENDER,
    RECEIVER,
    COMMON
}

data class Message (
    var participant: String?,
    var text: String,
    val id: UUID = UUID.randomUUID(),
    var contentType: String,
    var messageType: MessageType,
    var timeStamp: String? = null,
    var messageID: String? = null,
    var status: String? = null,
    var isRead: Boolean = false
){
    val content: MessageContent?
        get() = when (contentType) {
            ContentType.PLAIN_TEXT.type -> PlainTextContent.decode(text)
            ContentType.RICH_TEXT.type -> PlainTextContent.decode(text) // You can replace this with a rich text class later
            ContentType.INTERACTIVE_TEXT.type -> decodeInteractiveContent(text)
            else -> null // Handle unsupported content types
        }

    // Helper method to decode interactive content
    private fun decodeInteractiveContent(text: String): InteractiveContent? {
        val jsonData = text.toByteArray(Charsets.UTF_8)
        val genericTemplate = try {
            Json { ignoreUnknownKeys = true }.decodeFromString<GenericInteractiveTemplate>(String(jsonData))
        } catch (e: SerializationException) {
            null
        }
        return when (genericTemplate?.templateType) {
            QuickReplyContent.templateType -> QuickReplyContent.decode(text)
            ListPickerContent.templateType -> ListPickerContent.decode(text)
            // Add cases for each interactive message type, decoding as appropriate.
            else -> {
                println("Unsupported interactive content type: ${genericTemplate?.templateType}")
                null
            }
        }
    }

}